persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/StargateQueryHandler.java

```diff
@@ -102,4 +142,214 @@ public ResultMessage processBatch(
+        } else if (statement instanceof DropKeyspaceStatement) {
+          scope = Scope.DELETE;
+
+          DropKeyspaceStatement stmt = (DropKeyspaceStatement) statement;
+          keyspaceName = getKeyspaceName(stmt);
+        } else if (statement instanceof AlterKeyspaceStatement) {
+          scope = Scope.ALTER;
+
+          AlterKeyspaceStatement stmt = (AlterKeyspaceStatement) statement;
+          keyspaceName = getKeyspaceName(stmt);
+        }
+
+        logger.debug(
+            "preparing to authorize statement of type {} on {}.{}",
+            castStatement.getClass().toString(),
+            keyspaceName,
+            tableName);
+
+        try {
+          authorization.authorizeSchemaWrite(authToken, keyspaceName, tableName, scope);
+        } catch (io.stargate.auth.UnauthorizedException e) {
+          throw new UnauthorizedException(
+              String.format(
+                  "Missing correct permission on %s.%s",
+                  keyspaceName, (tableName == null ? "" : tableName)));
+        }
+
+        logger.debug(
+            "authorized statement of type {} on {}.{}",
+            castStatement.getClass().toString(),
+            keyspaceName,
+            tableName);
+      }
+    } finally {
+      if (authorization != null) {
+        ungetAuthorizationService();
+      }
+    }
+  }
+
+  private String getTableName(Object stmt) {
+    try {
+      Field f = stmt.getClass().getDeclaredField("tableName");
+      f.setAccessible(true);
+      return (String) f.get(stmt);
+    } catch (Exception e) {
+      logger.error("Unable to get private field", e);
+      throw new RuntimeException("Unable to get private field", e);
+    }
+  }
+
+  private String getKeyspaceName(Object stmt) {
+    try {
+      Field f = stmt.getClass().getSuperclass().getDeclaredField("keyspaceName");
+      f.setAccessible(true);
+      return (String) f.get(stmt);
+    } catch (Exception e) {
+      logger.error("Unable to get private field", e);
+      throw new RuntimeException("Unable to get private field", e);
+    }
+  }
+
+  private AuthorizationService getAuthorizationServiceFromContext() {
+    BundleContext bundleContext = FrameworkUtil.getBundle(this.getClass()).getBundleContext();
+
+    ServiceReference<?>[] refs;
+    try {
+      refs = bundleContext.getServiceReferences(AuthorizationService.class.getName(), null);
+    } catch (InvalidSyntaxException e) {
+      logger.error("Failed to get service reference for AuthorizationService", e);
+      return null;
+    }
+
+    if (refs != null) {
+      for (ServiceReference<?> ref : refs) {
+        Object service = bundleContext.getService(ref);
+        if (service instanceof AuthorizationService
+            && ref.getProperty("AuthIdentifier") != null
+            && ref.getProperty("AuthIdentifier").equals(System.getProperty("stargate.auth_id"))) {
+          authorizationReference = ref;
+          return (AuthorizationService) service;
+        }
```

Comment:
I believe we also need to "unget" when the `if` does not match.

Reply:
done